### PR TITLE
Fix issues in ImageDownload and ImageDownloader

### LINKED ISSUE #3067: Crash in ImageDownloader
**Mapbox Navigation SDK version:** 2.0.0-beta.12

### Steps to reproduce

```
Crashed: com.apple.NSURLSession-delegate
0  OneTap                         0x1007e72b0 ImageDownloader.activeOperation(with:) + 72 (ImageDownloader.swift:72)
1  OneTap                         0x1007e8530 specialized ImageDownloader.urlSession(_:dataTask:didReceive:completionHandler:) + 98 (<compiler-generated>:98)
2  OneTap                         0x1007e7408 @objc ImageDownloader.urlSession(_:dataTask:didReceive:completionHandler:) + 4339856392 (<compiler-generated>:4339856392)
3  CFNetwork                      0x199c15038 _CFNetworkHTTPConnectionCacheSetLimit + 155128
```

I have no exact steps to reproduce. The crash happened during a simulated navigation session here:

https://github.com/mapbox/mapbox-navigation-ios/blob/48c312c3fd2830a08b2079a58db535dfcaa8508e/Sources/MapboxNavigation/ImageDownloader.swift#L72

due to no synchronization in this method for `activeOperation` call:

https://github.com/mapbox/mapbox-navigation-ios/blob/48c312c3fd2830a08b2079a58db535dfcaa8508e/Sources/MapboxNavigation/ImageDownloader.swift#L97-L103

All 3 `URLSessionDataDelegate` methods should have synchronization for `operations` dictionary.

cc: @truburt @1ec5 @MaximAlien

### PR DESCRIPTION
**Pull request by S2Ler**
Fixes #3067

- Fix a lot of thread-safety issues.
- Fix incorrect handling of errors:
   - Completions are called two times for HTTP status code >= 400.
   - `ImageDownload` operation wasn't properly transition into finished
      state.
- `ImageDownload` operation didn't have proper `isReady` state.
- Increase test coverage for these types.

#### `ImageDownloader`
- `ImageDownloader` is fixed by properly using Dispatch Queue: perform protected state read/write inside the queue.
-  concurrent `accessQueue` replaced by a serial one, because most accesses to the queue modifies the protected state that require a barrier to be placed. That makes the `accessQueue` merely a serial queue. 
#### `ImageDownload`
`ImageDownload` is a subclass of NSOperation. 
- The implementation is improved with a simpler to use `State` variable. No more multiple flags juggling. 
- A queue replaced with a lock. 
- Replace `objc_sync_enter` with a lock.

### CODE REVIEW

File: Tests/MapboxNavigationTests/ImageDownloaderTests.swift [hunk 1]
```diff
@@ -135,4 +124,134 @@ class ImageDownloaderTests: XCTestCase {
...
+        XCTAssertNotNil(errorReturned)
+    }
+
+    func testDownloadWith400StatusCode() {
+        var imageReturned: UIImage?
+        var dataReturned: Data?
+        var errorReturned: Error?
+        let imageDownloaded = expectation(description: "Image Downloaded")
+
+        let faultyUrl = URL(string: "https://www.mapbox.com/incorrect")!
```
Comment (S2Ler) [inline Tests/MapboxNavigationTests/ImageDownloaderTests.swift:154]: Not very reliable way of testing 4xx status code. Will extend the url spy with ability to emulate errors

File: Sources/MapboxNavigation/ImageDownloader.swift [hunk 2]
```diff
@@ -11,47 +11,46 @@ protocol ReentrantImageDownloader {
...
-            self.operationType = op
-        }
+        self.accessQueue = DispatchQueue(label: Bundle.mapboxNavigation.bundleIdentifier! + ".ImageDownloaderInternal")
+        super.init()
+
+        // NOTE: `lazy var` isn't thread safe while `init` methods are.
+        //       `lazy var` for `urlSession` is used to specify `self` as a delegate which only available after
+        //       all properties are initialized ans `super.init` is called if appropriate.
+        //       So we trigger `init` here so that `urlSession` is initialized in a thread safe manner.
+        _ = urlSession
```
Comment (1ec5) [inline Sources/MapboxNavigation/ImageDownloader.swift:42]: Does the lazy syntax benefit us here? Maybe it’d be simpler to inline that initialization here.

**Review by 1ec5 — APPROVED** [review]
Locking internal to a class is always scary, but I think you’ve covered your bases.

File: Tests/MapboxNavigationTests/ImageDownloaderTests.swift (hunk 1 above)
Comment (S2Ler) [reply Tests/MapboxNavigationTests/ImageDownloaderTests.swift]: Done

File: Sources/MapboxNavigation/ImageDownloader.swift (hunk 2 above)
Comment (S2Ler) [reply Sources/MapboxNavigation/ImageDownloader.swift:42]: Good questions, thanks for asking. Without going deep into explaining the rationale why I've kept `lazy var` I think we can change `private lazy var urlSession: URLSession` to `private var urlSession: URLSession!`.

**Comment by S2Ler** [comment]
> Locking internal to a class is always scary, but I think you’ve covered your bases.

Sorry, can you please rephrase?

**Pull request merged**